File: typescript/src/executor.ts
/**
 * Executor — runs an ExecutionPlan with maximum parallelism.
 *
 * Features:
 * - Promise.all per Stage for true concurrent execution.
 * - Per-call timeout via AbortController + Promise.race.
 * - Retry with exponential back-off.
 * - Concurrency cap via a simple semaphore.
 * - $results.<id>[.<path>] placeholder substitution between stages.
 * - Rich ExecutionReport with per-call timing and speedup estimate.
 */

import type { ToolCall } from "./dependencyAnalyzer.js";
import type { OTelIntegration, SynapseLogger } from "./observability.js";
import type { ExecutionPlan } from "./planner.js";

export type AsyncToolFn = (...args: unknown[]) => Promise<unknown>;

/** Result for a single tool call. */
export interface CallResult {
  callId: string;
  toolName: string;
  status: "success" | "error" | "timeout";
  output?: unknown;
  error?: string;
  durationMs: number;
  attempts: number;
}

/** Full report returned by Executor.execute(). */
export interface ExecutionReport {
  results: Map<string, CallResult>;
  totalDurationMs: number;
  stagesRun: number;
  parallelCalls: number;
  sequentialCalls: number;
  /** sum(individual durations) / wall clock — ≥ 1 means parallelism saved time. */
  speedupEstimate: number;
}

// ── Placeholder resolution ─────────────────────────────────────────────────────

const REF_FULL = /^\$results\.([a-zA-Z0-9_\-.\[\]]+)$/;
const REF_INLINE = /\$results\.([a-zA-Z0-9_\-.\[\]]+)/g;

function splitRef(rawRef: string, results: Map<string, unknown>): [string, string | undefined] {
  const parts = rawRef.split(".");
  for (let i = parts.length; i > 0; i--) {
    const candidate = parts.slice(0, i).join(".");
    if (results.has(candidate)) {
      const suffix = parts.slice(i).join(".");
      return [candidate, suffix || undefined];
    }
  }
  return [rawRef, undefined];
}

function getNestedValue(obj: unknown, path: string | undefined): unknown {
  if (!path || obj === undefined || obj === null) return obj;
  for (const part of path.split(".")) {
    const arrayMatch = part.match(/^(.+)\[(\d+)\]$/);
    if (arrayMatch) {
      obj = (obj as Record<string, unknown>)[arrayMatch[1]];
      obj = (obj as unknown[])[parseInt(arrayMatch[2], 10)];
    } else if (typeof obj === "object" && obj !== null) {
      obj = (obj as Record<string, unknown>)[part];
    } else {
      return undefined;
    }
    if (obj === undefined) return undefined;
  }
  return obj;
}

function resolve(value: unknown, results: Map<string, unknown>): unknown {
  if (typeof value === "string") {
    const fullMatch = REF_FULL.exec(value);
    if (fullMatch) {
      const [id, path] = splitRef(fullMatch[1], results);
      return getNestedValue(results.get(id), path);
    }
    // Inline substitution (always produces a string)
    REF_INLINE.lastIndex = 0;
    return value.replace(REF_INLINE, (match, rawRef: string) => {
      const [id, path] = splitRef(rawRef, results);
      const v = getNestedValue(results.get(id), path);
      return v !== undefined ? String(v) : match;
    });
  }
  if (Array.isArray(value)) return value.map((item) => resolve(item, results));
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).map(([k, v]) => [
        k,
        resolve(v, results),
      ])
    );
  }
  return value;
}

// ── Semaphore ──────────────────────────────────────────────────────────────────

class Semaphore {
  private _count: number;
  private _queue: Array<() => void> = [];

  constructor(max: number) {
    this._count = max;
  }

  async acquire(): Promise<void> {
    if (this._count > 0) {
      this._count--;
      return;
    }
    return new Promise((resolve) => this._queue.push(resolve));
  }

  release(): void {
    if (this._queue.length > 0) {
      this._queue.shift()!();
    } else {
      this._count++;
    }
  }
}

// ── Executor ───────────────────────────────────────────────────────────────────

export interface ExecutorOptions {
  maxConcurrency?: number;
  defaultTimeoutMs?: number;
  defaultRetries?: number;
  onCallStart?: (call: ToolCall) => void;
  onCallEnd?: (result: CallResult) => void;
  logger?: SynapseLogger;
  otel?: OTelIntegration;
}

export class Executor {
  private tools: Map<string, AsyncToolFn>;
  private semaphore: Semaphore;
  private defaultTimeoutMs: number;
  private defaultRetries: number;
  private onCallStart: ((call: ToolCall) => void) | undefined;
  private onCallEnd: ((result: CallResult) => void) | undefined;
  private logger: SynapseLogger | undefined;
  private otel: OTelIntegration | undefined;

  constructor(tools: Record<string, AsyncToolFn>, options: ExecutorOptions = {}) {
    this.tools = new Map(Object.entries(tools));
    this.semaphore = new Semaphore(options.maxConcurrency ?? 16);
    this.defaultTimeoutMs = options.defaultTimeoutMs ?? 30_000;
    this.defaultRetries = options.defaultRetries ?? 0;
    this.onCallStart = options.onCallStart;
    this.onCallEnd = options.onCallEnd;
    this.logger = options.logger;
    this.otel = options.otel;
  }

  async execute(plan: ExecutionPlan): Promise<ExecutionReport> {
    const accumulated = new Map<string, unknown>();
    const allResults = new Map<string, CallResult>();
    let sumDurations = 0;
    let parallelCalls = 0;
    let sequentialCalls = 0;

    const wallStart = performance.now();

    for (const stage of plan.stages) {
      const stageStart = performance.now();
      this.logger?.log("stage_started", {
        stageIndex: stage.index,
        parallelCount: stage.calls.length,
        toolCount: stage.calls.length,
      });
      const stageSpan = this.otel?.startSpan("synapse.stage", {
        stageIndex: stage.index,
        toolCount: stage.calls.length,
      });
      if (stage.calls.length > 1) {
        parallelCalls += stage.calls.length;
      } else {
        sequentialCalls += stage.calls.length;
      }

      const stageResults = await Promise.all(
        stage.calls.map((call) => this.runCall(call, accumulated, stage.index, stage.calls.length))
      );

      const stageLatencyMs = performance.now() - stageStart;
      this.logger?.log("stage_completed", {
        stageIndex: stage.index,
        latencyMs: stageLatencyMs,
        parallelCount: stage.calls.length,
        toolCount: stage.calls.length,
      });
      this.otel?.endSpan(stageSpan, stageResults.every((result) => result.status === "success"));

      for (const result of stageResults) {
        allResults.set(result.callId, result);
        accumulated.set(result.callId, result.output);
        sumDurations += result.durationMs;
      }
    }

    const totalDurationMs = performance.now() - wallStart;
    const speedupEstimate = totalDurationMs > 0 ? sumDurations / totalDurationMs : 1;
    const avgParallelism = plan.stages.length > 0 ? plan.totalCalls / plan.stages.length : 0;
    this.logger?.recordExecution(speedupEstimate, avgParallelism);
    this.logger?.log("execution_complete", {
      latencyMs: totalDurationMs,
      parallelCount: parallelCalls,
      totalCalls: plan.totalCalls,
      speedupRatio: speedupEstimate,
    });

    return {
      results: allResults,
      totalDurationMs,
      stagesRun: plan.stages.length,
      parallelCalls,
      sequentialCalls,
      speedupEstimate,
    };
  }

  private async runCall(
    call: ToolCall,
    accumulated: Map<string, unknown>,
    stageIndex: number,
    parallelCount: number
  ): Promise<CallResult> {
    const toolFn = this.tools.get(call.name);
    if (!toolFn) {
      this.logger?.log("tool_failed", {
        stageIndex,
        toolId: call.id,
        parallelCount,
        error: `No tool registered for '${call.name}'`,
      });
      return {
        callId: call.id,
        toolName: call.name,
        status: "error",
        error: `No tool registered for '${call.name}'`,
        durationMs: 0,
        attempts: 0,
      };
    }

    const resolvedInputs = resolve(call.inputs ?? {}, accumulated) as Record<
      string,
      unknown
    >;
    const timeoutMs = call.timeoutMs ?? this.defaultTimeoutMs;
    const maxAttempts = 1 + (call.retries ?? this.defaultRetries);

    let lastError = "";
    let lastStatus: "error" | "timeout" = "error";
    let lastDuration = 0;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      this.onCallStart?.(call);
      this.logger?.log("tool_started", {
        stageIndex,
        toolId: call.id,
        parallelCount,
      });
      const toolSpan = this.otel?.startSpan("synapse.tool_call", {
        stageIndex,
        toolId: call.id,
        toolName: call.name,
        attempt,
      });
      const t0 = performance.now();

      try {
        await this.semaphore.acquire();
        let output: unknown;
        try {
          output = await Promise.race([
            toolFn(...Object.values(resolvedInputs)),
            new Promise<never>((_, reject) =>
              setTimeout(() => reject(new Error(`__timeout__`)), timeoutMs)
            ),
          ]);
        } finally {
          this.semaphore.release();
        }

        const durationMs = performance.now() - t0;
        const result: CallResult = {
          callId: call.id,
          toolName: call.name,
          status: "success",
          output,
          durationMs,
          attempts: attempt,
        };
        this.onCallEnd?.(result);
        this.logger?.recordToolResult(durationMs, false);
        this.logger?.log("tool_completed", {
          stageIndex,
          toolId: call.id,
          latencyMs: durationMs,
          parallelCount,
        });
        this.otel?.endSpan(toolSpan, true);
        return result;
      } catch (err: unknown) {
        lastDuration = performance.now() - t0;
        const msg = err instanceof Error ? err.message : String(err);
        if (msg === "__timeout__") {
          lastStatus = "timeout";
          lastError = `Timed out after ${timeoutMs}ms`;
        } else {
          lastStatus = "error";
          lastError = msg;
        }
        this.otel?.endSpan(toolSpan, false, lastError);

        if (attempt < maxAttempts) {
          await new Promise((r) => setTimeout(r, 100 * 2 ** (attempt - 1)));
        }
      }
    }

    const result: CallResult = {
      callId: call.id,
      toolName: call.name,
      status: lastStatus,
      error: lastError,
      durationMs: lastDuration,
      attempts: maxAttempts,
    };
    this.onCallEnd?.(result);
    this.logger?.recordToolResult(lastDuration, true);
    this.logger?.log("tool_failed", {
      stageIndex,
      toolId: call.id,
      latencyMs: lastDuration,
      parallelCount,
      error: lastError,
      status: lastStatus,
    });
    return result;
  }
}

File: python/synapse/executor.py
"""
Executor — runs an ExecutionPlan with async concurrency.

Features
--------
* Fully async via asyncio.gather — zero threads needed.
* Per-call timeout and retry with exponential back-off.
* Concurrency cap via asyncio.Semaphore (don't overwhelm rate-limited APIs).
* $results.<id> placeholder substitution so later stages receive real values.
* Rich execution report with per-call timing and status.
"""

from __future__ import annotations

import asyncio
import re
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Coroutine

from .dependency_analyzer import ToolCall
from .planner import ExecutionPlan

# ── Type alias for an async tool implementation ───────────────────────────────
AsyncToolFn = Callable[..., Coroutine[Any, Any, Any]]

_REF_PATTERN = re.compile(r"\$results\.([a-zA-Z0-9_\-\.\[\]]+)")


# ── Per-call result ────────────────────────────────────────────────────────────

@dataclass
class CallResult:
    call_id: str
    tool_name: str
    status: str          # "success" | "error" | "timeout"
    output: Any = None
    error: str | None = None
    duration_ms: float = 0.0
    attempts: int = 1


# ── Full execution report ──────────────────────────────────────────────────────

@dataclass
class ExecutionReport:
    results: dict[str, CallResult]
    total_duration_ms: float
    stages_run: int
    parallel_calls: int   # calls that ran concurrently with at least one other
    sequential_calls: int
    speedup_estimate: float  # sum(durations) / wall_clock (≥1 means parallelism helped)

    def __repr__(self) -> str:
        lines = [
            "── Synapse Execution Report ──────────────────────────────────",
            f"  Wall clock : {self.total_duration_ms:.1f} ms",
            f"  Stages     : {self.stages_run}",
            f"  Parallel   : {self.parallel_calls} calls",
            f"  Sequential : {self.sequential_calls} calls",
            f"  Speedup    : {self.speedup_estimate:.2f}x",
            "  Results:",
        ]
        for r in self.results.values():
            icon = "✓" if r.status == "success" else "✗"
            lines.append(
                f"    {icon} [{r.call_id}] {r.tool_name} "
                f"— {r.status} in {r.duration_ms:.1f} ms"
                + (f" (attempt {r.attempts})" if r.attempts > 1 else "")
            )
        lines.append("──────────────────────────────────────────────────────────")
        return "\n".join(lines)


# ── Placeholder resolver ───────────────────────────────────────────────────────

def _resolve(value: Any, results: dict[str, Any]) -> Any:
    """Replace $results.<id>[.<path>] placeholders with actual values."""
    if isinstance(value, str):
        # Single full-match replacement preserves non-string types
        full = _REF_PATTERN.fullmatch(value)
        if full:
            ref_id, path = _split_ref(full.group(1), results)
            return _get_nested(results.get(ref_id), path)
        # Inline replacement (always produces a string)
        def replacer(m: re.Match[str]) -> str:
            ref_id, path = _split_ref(m.group(1), results)
            v = _get_nested(results.get(ref_id), path)
            return str(v) if v is not None else m.group(0)
        return _REF_PATTERN.sub(replacer, value)
    elif isinstance(value, dict):
        return {k: _resolve(v, results) for k, v in value.items()}
    elif isinstance(value, list):
        return [_resolve(item, results) for item in value]
    return value


def _split_ref(raw_ref: str, results: dict[str, Any]) -> tuple[str, str | None]:
    parts = raw_ref.split(".")
    for i in range(len(parts), 0, -1):
        candidate = ".".join(parts[:i])
        if candidate in results:
            suffix = ".".join(parts[i:])
            return candidate, (suffix or None)

    return raw_ref, None


def _get_nested(obj: Any, path: str | None) -> Any:
    if obj is None or path is None:
        return obj
    for part in path.split("."):
        if part.endswith("]"):
            key, _, idx = part.rstrip("]").partition("[")
            obj = obj[key][int(idx)]
        elif isinstance(obj, dict):
            obj = obj.get(part)
        else:
            obj = getattr(obj, part, None)
        if obj is None:
            break
    return obj


# ── Executor ───────────────────────────────────────────────────────────────────

class Executor:
    """
    Executes an ExecutionPlan against a registry of async tool functions.

    Parameters
    ----------
    tools:
        Mapping from tool name → async callable.
    max_concurrency:
        Maximum simultaneous in-flight tool calls.  Defaults to 16.
    default_timeout:
        Seconds before a call is cancelled if the ToolCall has no timeout.
    default_retries:
        Retry count if the ToolCall has no retries set.
    on_call_start / on_call_end:
        Optional hooks for observability / progress reporting.
    """

    def __init__(
        self,
        tools: dict[str, AsyncToolFn],
        *,
        max_concurrency: int = 16,
        default_timeout: float = 30.0,
        default_retries: int = 0,
        on_call_start: Callable[[ToolCall], None] | None = None,
        on_call_end: Callable[[CallResult], None] | None = None,
    ) -> None:
        self.tools = tools
        self._semaphore = asyncio.Semaphore(max_concurrency)
        self.default_timeout = default_timeout
        self.default_retries = default_retries
        self.on_call_start = on_call_start
        self.on_call_end = on_call_end

    async def execute(self, plan: ExecutionPlan) -> ExecutionReport:
        """Run the full plan and return a detailed ExecutionReport."""
        accumulated_results: dict[str, Any] = {}   # call_id → raw output
        call_results: dict[str, CallResult] = {}

        wall_start = time.perf_counter()
        sum_call_duration = 0.0
        parallel_calls = 0
        sequential_calls = 0

        for stage in plan.stages:
            is_parallel = len(stage.calls) > 1
            if is_parallel:
                parallel_calls += len(stage.calls)
            else:
                sequential_calls += len(stage.calls)

            stage_results = await asyncio.gather(
                *[
                    self._run_call(call, accumulated_results)
                    for call in stage.calls
                ],
                return_exceptions=False,
            )

            for result in stage_results:
                call_results[result.call_id] = result
                accumulated_results[result.call_id] = result.output
                sum_call_duration += result.duration_ms

        wall_ms = (time.perf_counter() - wall_start) * 1000
        speedup = sum_call_duration / wall_ms if wall_ms > 0 else 1.0

        return ExecutionReport(
            results=call_results,
            total_duration_ms=wall_ms,
            stages_run=len(plan.stages),
            parallel_calls=parallel_calls,
            sequential_calls=sequential_calls,
            speedup_estimate=speedup,
        )

    async def _run_call(
        self, call: ToolCall, accumulated_results: dict[str, Any]
    ) -> CallResult:
        tool_fn = self.tools.get(call.name)
        if tool_fn is None:
            return CallResult(
                call_id=call.id,
                tool_name=call.name,
                status="error",
                error=f"No tool registered for '{call.name}'",
            )

        resolved_inputs = _resolve(call.inputs, accumulated_results)
        timeout = call.timeout if call.timeout is not None else self.default_timeout
        max_attempts = 1 + (call.retries if call.retries else self.default_retries)

        last_error: str = ""
        last_status: str = "error"
        last_duration_ms: float = 0.0
        for attempt in range(1, max_attempts + 1):
            if self.on_call_start:
                self.on_call_start(call)
            t0 = time.perf_counter()
            try:
                async with self._semaphore:
                    output = await asyncio.wait_for(
                        tool_fn(**resolved_inputs), timeout=timeout
                    )
                duration_ms = (time.perf_counter() - t0) * 1000
                result = CallResult(
                    call_id=call.id,
                    tool_name=call.name,
                    status="success",
                    output=output,
                    duration_ms=duration_ms,
                    attempts=attempt,
                )
                if self.on_call_end:
                    self.on_call_end(result)
                return result
            except asyncio.TimeoutError:
                last_duration_ms = (time.perf_counter() - t0) * 1000
                last_error = f"Timed out after {timeout}s"
                last_status = "timeout"
            except Exception as exc:  # noqa: BLE001
                last_duration_ms = (time.perf_counter() - t0) * 1000
                last_error = str(exc)
                last_status = "error"

            if attempt < max_attempts:
                await asyncio.sleep(0.1 * (2 ** (attempt - 1)))  # back-off

        result = CallResult(
            call_id=call.id,
            tool_name=call.name,
            status=last_status,
            error=last_error,
            duration_ms=last_duration_ms,
            attempts=max_attempts,
        )
        if self.on_call_end:
            self.on_call_end(result)
        return result
